File: py/TwoSum.py
# Given an integer array nums and integer target, return indices of the two numbers such that they add up to target.
# You may assume that each input would have exactly one solution, and you may not use the same element twice.
# You can return the answer in any order.


class Solution:
    def TwoSum(self, nums, target):
        hash_map = {}  # Create an empty dictionary to store the elements and their indices
        for i in range(len(nums)):
            # Calculate the complement of the current element with the target
            complement = target - nums[i]
            if complement in hash_map: # Check if the complement is already present in the hash map
                return [hash_map[complement], i]
            hash_map[nums[i]] = i

s = Solution()
print(s.TwoSum(nums=[2, 7, 11, 15], target=9))



# output: [0, 1]

# input: [2, 7, 11, 15], 9
# output: [0, 1]


# Explanation of the solution
# We can solve this problem using a hash map to store the elements and their indices.
# We iterate over the input array and calculate the complement of the current element with the target value.
# If the complement is already present in the hash map, we return the indices of the two elements that add up to the target.
# If the complement is not present in the hash map, we add the current element to the hash map with its index.
# This way, we can find the indices of the two numbers that add up to the target value in O(n) time complexity.
# The space complexity of this solution is O(n) because we store the elements and their indices in the hash map.
# This solution is efficient and does not require sorting the input array.
# We can return the indices in any order as per the problem statement.
# The solution is implemented in the TwoSum method of the Solution class.

File: MCAJAVA/TwoSum.java
// Given an integer array nums and an integer target return the two indicies such that they add up to the target 
// you can only use one element at a time 
// Example : nums = [2, 7, 11, 15], target = 9
// output : [0, 1] because the two indices nums[0] = 2 + nums[1] = 7 gives the output: 9

import java.util.*;
class TwoSum{
    public int[] TwoSumPair(int[] nums, int target){
	int n = nums.length;
	var map = new HashMap<Integer, Integer>();
	for(int i = 0; i < n; i++){
		int compliment = target - nums[i];
		if (map.containsKey(compliment)){
			return new int[]{map.get(compliment), i};
		} 
		map.put(nums[i], i);
	}
	return new int[]{};
    }
    public static void main(String args[]){
	var ts = new TwoSum();
	int nums[] = {2, 7, 11, 15};
	int target = 9;
	System.out.println(Arrays.toString(ts.TwoSumPair(nums, target)));
    }
} 


// nums = [2, 7, 11, 15], target = 9
// brute force : check every pair available and return those indices that gives us the valid ouput
// map = {};
// 
// this approach gives us the time complexity of O(n^2)
